presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +153,135 @@ private boolean advanceToNextRowGroup()
     public Block readArray(Type type, List<String> path)
             throws IOException
     {
-        return readArray(type, path, new IntArrayList());
+        return readArray(type, path, new IntArrayList(), new IntArrayList());
     }
 
-    private Block readArray(Type type, List<String> path, IntList elementOffsets)
+    private Block readArray(Type type, List<String> path, IntList definitionLevels, IntList repetitionLevels)
             throws IOException
     {
         List<Type> parameters = type.getTypeParameters();
         checkArgument(parameters.size() == 1, "Arrays must have a single type parameter, found %d", parameters.size());
         path.add(ARRAY_TYPE_NAME);
         Type elementType = parameters.get(0);
-        Block block = readBlock(ARRAY_ELEMENT_NAME, elementType, path, elementOffsets);
-        path.remove(ARRAY_TYPE_NAME);
+        Block block = readBlock(ARRAY_ELEMENT_NAME, elementType, path, definitionLevels, repetitionLevels);
+        path.remove(path.lastIndexOf(ARRAY_TYPE_NAME));
+        IntList offsets = new IntArrayList();
+        BooleanList valueIsNull = new BooleanArrayList();
+        calculateOffsets(path, offsets, valueIsNull, definitionLevels, repetitionLevels);
+        return new ArrayBlock(valueIsNull.size(), valueIsNull.toBooleanArray(), offsets.toIntArray(), block);
+    }
 
-        if (elementOffsets.isEmpty()) {
-            for (int i = 0; i < batchSize; i++) {
-                elementOffsets.add(0);
+    private void calculateOffsets(List<String> path, IntList offsets, BooleanList valueIsNull, IntList definitionLevels, IntList repetitionLevels)
+    {
+        int valueOffset;
+        int maxDefinitionLevel = getMaxDefinitionLevel(path);
+        int maxRepetitionLevel = getMaxRepetitionLevel(path);
+        offsets.add(0);
+        for (int i = 0; i < definitionLevels.size(); i++) {
+            valueOffset = 1;
+            while (i + 1 < repetitionLevels.size() && repetitionLevels.get(i + 1) > maxRepetitionLevel) {
+                if (repetitionLevels.get(i + 1) <= maxRepetitionLevel + 1) {
+                    valueOffset++;
+                }
+                i++;
+            }
+            if (definitionLevels.get(i) == maxDefinitionLevel - 1) {
```

Comment:
this part of the code looks error prone, please add some comments/links to the relevant parts of the spec that explains what you do here.

Reply:
refactored